java/arcs/core/analysis/AbstractValue.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+/**
+ * An interface for defining the values of an abstract domain commonly used in abstract
+ * interpretation frameworks (https://en.wikipedia.org/wiki/Abstract_interpretation).
+ *
+ * An abstract domain is typically defined as a mathematical lattice, but they don't have to be a
+ * lattice in general. The interface here is designed to be general enough to allow abstract domains
+ * that are not lattices (https://en.wikipedia.org/wiki/Lattice_(order)). To that extent, we only
+ * expect the implementing class to provide an [isEquivalentTo] method that returns true if two
+ * values are semantically equivalent. The [isEquivalentTo] method is used to determine if we
+ * have reached a fixpoint when using iterative algorithms to solve the data flow equations for the
+ * problem at hand (https://en.wikipedia.org/wiki/Data-flow_analysis). For abstract domains that are
+ * lattices, the [isEquivalentTo] method can be used to compute the partial order as follows:
+ *    `a.isLessThanEqual(b)` <=> `b.isEquivalentTo(a.join(b))` [for join semi-lattice]
+ *    `a.isLessThanEqual(b)` <=> `a.isEquivalentTo(a.meet(b))` [for meet semi-lattice]
```

Comment:
Nit: I don't think KDoc has meaningful whitespace, so this should use bullet points

Reply:
Done.